vm/src/stdlib/array.rs

```diff
@@ -1,1067 +1,1418 @@
+                (true, false) => <$SIGNED_TY>::from_le_bytes(b).into_pyobject($VM),
+                (true, true) => <$SIGNED_TY>::from_be_bytes(b).into_pyobject($VM),
+            }
+        }};
+    }
 
-impl PyIter for PyArrayIter {
-    fn next(zelf: &PyRef<Self>, vm: &VirtualMachine) -> PyResult {
-        let pos = zelf.position.fetch_add(1);
-        if let Some(item) = zelf.array.read().getitem_by_idx(pos, vm)? {
-            Ok(item)
-        } else {
-            Err(vm.new_stop_iteration())
+    #[pyfunction]
+    fn _array_reconstructor(args: ReconstructorArgs, vm: &VirtualMachine) -> PyResult<PyArrayRef> {
+        let cls = check_array_type(args.arraytype, vm)?;
+        let mut array = check_type_code(args.typecode, vm)?;
+        let format = args.mformat_code;
+        let bytes = args.items.as_bytes();
+        if bytes.len() % format.item_size() != 0 {
+            return Err(vm.new_value_error("bytes length not a multiple of item size".into()));
+        }
+        if MachineFormatCode::from_typecode(array.typecode()) == Some(format) {
+            array.frombytes(bytes);
+            return PyArray::from(array).into_ref_with_type(vm, cls);
+        }
+        if !matches!(
+            format,
+            MachineFormatCode::Utf16 { .. } | MachineFormatCode::Utf32 { .. }
+        ) {
+            array.reserve(bytes.len() / format.item_size());
         }
+        let mut chunks = bytes.chunks(format.item_size());
+        match format {
+            MachineFormatCode::Ieee754Float { big_endian } => {
+                chunks.try_for_each(|b| array.push(chunk_to_obj!(vm, b, f32, big_endian), vm))?
+            }
+            MachineFormatCode::Ieee754Double { big_endian } => {
+                chunks.try_for_each(|b| array.push(chunk_to_obj!(vm, b, f64, big_endian), vm))?
+            }
+            MachineFormatCode::Int8 { signed } => chunks
+                .try_for_each(|b| array.push(chunk_to_obj!(vm, b, i8, u8, signed, false), vm))?,
+            MachineFormatCode::Int16 { signed, big_endian } => chunks.try_for_each(|b| {
+                array.push(chunk_to_obj!(vm, b, i16, u16, signed, big_endian), vm)
+            })?,
+            MachineFormatCode::Int32 { signed, big_endian } => chunks.try_for_each(|b| {
+                array.push(chunk_to_obj!(vm, b, i32, u32, signed, big_endian), vm)
+            })?,
+            MachineFormatCode::Int64 { signed, big_endian } => chunks.try_for_each(|b| {
+                array.push(chunk_to_obj!(vm, b, i64, u64, signed, big_endian), vm)
+            })?,
+            MachineFormatCode::Utf16 { big_endian } => {
+                let utf16: Vec<_> = chunks.map(|b| chunk_to_obj!(b, u16, big_endian)).collect();
+                let s = String::from_utf16(&utf16).map_err(|_| {
+                    vm.new_unicode_encode_error("items cannot decode as utf16".into())
+                })?;
+                eprintln!("utf16: {:?}", s);
```

Comment:
these `eprintln`s look like debug prints probably not inteded to be included in final PR